### DISCUSSION TITLE
Usage of PERSISTENT fields to speed up SQL

### DISCUSSION THREAD

**Discussion by pakricard** [timschofield/webERP#238] [Development Discussion & Specification]
Hi all:

This subject popped up on another thread, but I guess it deserves a new discussion thread.

To speed up several SQL queries, I added some persistent fields in:
```
ALTER TABLE `stockmaster` ADD `actualcost` decimal(20,4) AS (materialcost+labourcost+overheadcost) PERSISTENT;
ALTER TABLE `debtortrans` ADD `balance` double AS (ovamount + ovgst + ovfreight + ovdiscount - alloc) PERSISTENT;
ALTER TABLE `loctransfers` ADD `pendingqty` double AS (shipqty-recqty) PERSISTENT;
ALTER TABLE `salesorderdetails` ADD ``linenetprice` double AS (qtyinvoiced * (unitprice * (1 - discountpercent))) PERSISTENT;
```

The idea behind is to modify all SQL queries in webERP using the AS expression, replacing it by the calculated field. 

Tried some others but did not speed up webERP much, but it might depend on each installation, table size, query result size, etc.

Gemini summarized this as:

Generated Columns (PERSISTENT and VIRTUAL):
Advantage:
- Simplify Queries: By calculating values on-the-fly, you can reduce the complexity of queries, especially those involving complex expressions.   
- Improve Query Performance: For frequently accessed calculated values, persistent generated columns can improve performance by pre-calculating and storing the values.
- Data Integrity: Ensures data consistency by automatically updating dependent columns when the underlying data changes.

Types:
- PERSISTENT: The value is calculated and stored in the database, making it accessible for indexing and querying.   
- VIRTUAL: The value is calculated on-the-fly when the column is accessed, providing flexibility but potentially impacting performance for complex calculations.

**Comment by timschofield** [comment]
Hi Ricard, I am going to have to read more on this as I can't get it to
work on my machine here. Are you using mysql or mariadb? What is your
sql-mode set to?

Thanks
Tim

On Fri, 15 Nov 2024 at 01:57, pakricard ***@***.***> wrote:

> Hi all:
>
> This subject popped up on another thread, but I guess it deserves a new
> discussion thread.
>
> To speed up several SQL queries, I added some persistent fields in:
>
> ALTER TABLE `stockmaster` ADD `actualcost` decimal(20,4) AS (materialcost+labourcost+overheadcost) PERSISTENT;
> ALTER TABLE `debtortrans` ADD `balance` double AS (ovamount + ovgst + ovfreight + ovdiscount - alloc) PERSISTENT;
> ALTER TABLE `loctransfers` ADD `pendingqty` double AS (shipqty-recqty) PERSISTENT;
> ALTER TABLE `salesorderdetails` ADD ` `linenetprice` double AS (qtyinvoiced * (unitprice * (1 - discountpercent))) PERSISTENT;
>
> The idea behind is to modify all SQL queries in webERP using the AS
> expression, replacing it by the calculated field.
>
> Tried some others but did not speed up webERP much, but it might depend on
> each installation, table size, query result size, etc.
>
> Gemini summarized this as:
>
> Generated Columns (PERSISTENT and VIRTUAL):
> Advantage:
>
>    - Simplify Queries: By calculating values on-the-fly, you can reduce
>    the complexity of queries, especially those involving complex expressions.
>    - Improve Query Performance: For frequently accessed calculated
>    values, persistent generated columns can improve performance by
>    pre-calculating and storing the values.
>    - Data Integrity: Ensures data consistency by automatically updating
>    dependent columns when the underlying data changes.
>
> Types:
>
>    - PERSISTENT: The value is calculated and stored in the database,
>    making it accessible for indexing and querying.
>    - VIRTUAL: The value is calculated on-the-fly when the column is
>    accessed, providing flexibility but potentially impacting performance for
>    complex calculations.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/timschofield/webERP/discussions/238>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAL6LEFM2MUTEMEOZE5MFT2AVIIHAVCNFSM6AAAAABR2F5IMSVHI2DSMVQWIX3LMV43ERDJONRXK43TNFXW4OZXGQ4DKNBZGE>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
www.weberp.org <http://www.weberpafrica.com>
@TimSchofield2
Blog: https://kwamoja.home.blog/

**Comment by pakricard** [comment]
Hi Tim:
I am using MariaDB 5.5.68 under a VPS CentOS Linux 7.9.2009 (Core) with Plesk Obsidian 18.0.65, with no specific SQL mode.

![image](https://github.com/user-attachments/assets/3d1116a0-e60d-4646-8707-07c6548f4b9a)

Once I am sure my webERP is PHP 8 compatible, will be moved to a new VPS running Almalinux 9 and MariaDB 10.6

**Comment by timschofield** [comment]
Reading into this a bit more I think persistent columns are only in
MariaDB. MySql has stored columns which are similar. This brings up
the subject of whether we should have a ConnectDB_mariadb.inc file as
well as a ConnectDB_mysql.inc file as they appear to have started to
diverge?

Thanks
Tim

On Mon, 25 Nov 2024 at 08:03, pakricard ***@***.***> wrote:
>
> Hi Tim:
> I am using MariaDB 5.5.68 under a VPS CentOS Linux 7.9.2009 (Core) with Plesk Obsidian 18.0.65, with no specific SQL mode.
>
> image.png (view on web)
>
> Once I am sure my webERP is PHP 8 compatible, will be moved to a new VPS running Almalinux 9 and MariaDB 10.6
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you commented.Message ID: ***@***.***>



-- 
www.weberp.org
@TimSchofield2
Blog: https://kwamoja.home.blog/

**Comment by pakricard** [comment]
I found: https://dev.mysql.com/doc/refman/8.4/en/create-table-generated-columns.html , so it seems MySQL has a similar application, but the syntax is different.

The PERSISTENT in MariaDB is called STORED in MySQL

ConnectDB_mariadb.inc? Probably, depending on how many users are using MariaDB . My VPS supplier changed a few years ago from MySQL to MariaDB, and never faced any issue, I was (wrongly) sure they were synonyms.

Not sure how popularMariaDB is or is just me amongst webERP users?

Thanks,
R

**Comment by dalers** [comment]
I prefer MariaDb vs Oracle MySQL for their more open project philosophy.

**Comment by timschofield** [comment]
As they do appear to be diverging (if only slowly for now) I don't see that
we have much alternative but to create a ConnectDB_mariadb.inc script. We
can't really mandate one as some people are tied into their hosting
companies choice. We would then need to offer that choice in the installer.

Do you agree?
Tim

On Tue, 26 Nov 2024 at 06:32, Dale Scott ***@***.***> wrote:

> I prefer MariaDb because it's a fully open source project.
>
> From MS Copilot:
>
> MariaDB and MySQL are both popular relational database management
> systems (RDBMS), but they have some key differences:
>
> 1. **Origin**:
> - **MySQL**: Originally developed by MySQL AB, it was acquired by
> Oracle Corporation.
> - **MariaDB**: A fork of MySQL, created by the original developers of
> MySQL to ensure the project remains open-source.
>
> 2. **Licensing**:
> - **MySQL**: Uses a dual-license approach, offering both open-source
> (GPL) and proprietary licenses.
> - **MariaDB**: Fully open-source under the GPL license.
>
> 3. **Development and Features**:
> - **MySQL**: Development is controlled by Oracle, which has
> introduced some closed-source modules.
> - **MariaDB**: Maintains compatibility with MySQL but often includes
> additional features and improvements, such as new storage engines and
> performance enhancements.
>
> 4. **Performance**:
> - **MariaDB**: Generally considered to have better performance, with
> improvements in speed and efficiency.
> - **MySQL**: Performance can vary, but it remains a robust and
> widely-used database system.
>
> 5. **Compatibility**:
> - **MariaDB**: Designed to be a drop-in replacement for MySQL,
> meaning it can use the same data and table definition files, client
> protocols, and APIs.
> - **MySQL**: While it remains compatible with many systems, some
> users prefer MariaDB to avoid potential vendor lock-in with Oracle.
>
> 6. **Community and Support**:
> - **MySQL**: Backed by Oracle, with a large user base and extensive
> documentation.
> - **MariaDB**: Supported by a strong open-source community, including
> many of the original MySQL developers.
>
> Both databases are powerful and capable, but your choice might depend on
> your specific needs, such as licensing preferences, performance
> requirements, and community support.
>
> On 2024-11-25 23:21, pakricard wrote:
>
> > I found:
> >
> https://dev.mysql.com/doc/refman/8.4/en/create-table-generated-columns.html
> > , so it seems MySQL has a similar application, but the syntax is
> > different.
> >
> > The PERSISTENT in MariaDB is called STORED in MySQL
> >
> > ConnectDB_mariadb.inc? Probably, depending on how many users are using
> > MariaDB . My VPS supplier changed a few years ago from MySQL to
> > MariaDB, and never faced any issue, I was (wrongly) sure they were
> > synonyms.
> >
> > Not sure how popularMariaDB is or is just me amongst webERP users?
> >
> > Thanks,
> > R
> >
> > --
> > Reply to this email directly, view it on GitHub [1], or unsubscribe
> > [2].
> > You are receiving this because you are subscribed to this
> > thread.Message ID:
> > ***@***.***>
>
>
> Links:
> ------
> [1]
>
> https://github.com/timschofield/webERP/discussions/238#discussioncomment-11380094
> [2]
>
> https://github.com/notifications/unsubscribe-auth/AAKFYYD4AJAFGOB7XHXS74T2CQHQHAVCNFSM6AAAAABR2F5IMSVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMZYGAYDSNA
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/timschofield/webERP/discussions/238#discussioncomment-11380136>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAL6LFBEYVGSV4E2MVC5XD2CQIWNAVCNFSM6AAAAABR2F5IMSVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMZYGAYTGNQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
www.weberp.org <http://www.weberpafrica.com>
@TimSchofield2
Blog: https://kwamoja.home.blog/

**Comment by dalers** [comment]
How about this as a solution - keep the master branch both MySQL _and_ MariaDb compatible and create a new branch for optimising either MySQL or MariaDb depending which has the greatest value.

Would it be possible to code a simple abstraction that would support either? It would make it easier to support the other dbms later with equivalent optimisations, or could possibly even be abandoned. 

Although my vote is still for MariaDb (and I think would be of many who get the choice), webERP is really for users and _their_ hosting arrangements, and probably everyone else knows more about a real webERP user than me. However, imho it seems unlikely companies for which webERP is a good fit will eventually migrate to commercial Oracle. It seems more likely they will either be acquired, in which case they will be merged into an existing more advanced system, or their needs will continue to be met by webERP - the lowest cost best solution for their needs all things considered. Not considering the Oracle advantage, it seems "open source" and possibly performance are on the side of MariaDb, and commercial hosting on the side of MySQL. 

Is there a sense for how webERP is hosted? 

I checked Digital Ocean, who I have used in the past but am not currently. Besidews standard Linux VPS, they provide managed databases - but MySQL not MariaDb (I just checked). I'm not surprised as I cynically suspect Oracle may be compensating Digital Ocean to only offer MySQL (which would really be no different than many other cross-licensing agreements, although I have no reason to believe it).

https://www.digitalocean.com/products/managed-databases?utm_campaign=search_us_en_brand&utm_adgroup=databases&utm_keyword=digitalocean%20dbaas&utm_matchtype=e&utm_adposition=&utm_creative=713874993326&utm_placement=&utm_device=c&utm_location=&utm_location=9001329&utm_term=digitalocean%20dbaas&utm_source=google&utm_medium=cpc&gad_source=1&gclid=Cj0KCQiAgJa6BhCOARIsAMiL7V93O55ficwotLkFgicycQuvEEu11kLesT3ChlZezhByzCHiqnL6ErcaAhZEEALw_wcB

I also checked WHC (World Hosting Canada, one of my IP registrars and DNS hosting providers, the other is No-IP), I know, you've never heard of them, but they also provide VPS (https://whc.ca/self-managed-cloud-servers) and support Softaculus for one-click installs (iiuc, I've never used nor understand Softaculus). The Softaculus library includes webERP 4.15.2 but neither MySQL or MariaDB.

In either case I would have to ssh login to a unix VPS and install everything myself, which is what I suspect many (most?) current webERP users are doing.

However, I assume Digital Ocean is not the only managed database provider. Is there any reason why one would use a managed database with webERP? I guess if you had a Softaculus one-click install of webERP on a WHC VPS, and a one-click creation of a managed database on Digital Ocean, and you could configure webERP to use the managed database, you wouldn't need a systems administrator....

@pakricard @efeoli who are your hosting service providers? @timschofield how are your clients hosting webERP?

Thoughts?

Fwiw, the Softaculus page for webERP 4.15.2 ([https://softaculous.com/apps/erp](https://softaculous.com/apps/erp)) links to [http://www.weberp.org/Support.html](http://www.weberp.org/Support.html) which doesn't exist.

**Comment by dalers** [comment]
P.S. if "most" webERP users can install whatever dbms they want, perhaps the question isn't mySQL or MariaDb, it's whether to resurrect the PostgreSQL adapter. PostgreSQL does seem to have become the defacto _professional_ grade open source dbms...

**Comment by dalers** [comment]
P.P.S. I see now Digital Ocean supports Plesk on their VPS (free for up to three domains). Iiuc correctly @pakricard you are using Plesk, how does it work for you? It seems Digital Ocean using Plesk has only MariaDb, although Digital Ocean seperately offers PostgreSQL managed databases (just but not a MariaDb managed database).

**Comment by timschofield** [comment]
Hi Dale, we do have that layer of abstraction with the ConnectDB_XXX.inc
and the UpdateDB_XXX.php files, where XXX is the DBMS being used.

So in the example of the calculated fields, both UpdateDB_mysqli.php and
UpdateDB_mariadb.php would have a function called something like
CreateCalculatedField(). In the mariadb file it would create a persistent
field, and in the mysqli file it would create a stored field. The rest of
webERP would just see that field and not care about which it was.

I would like to create postgres files as well, but I'm pretty sure there
aren't enough hours in the day :)

Tim

On Tue, 26 Nov 2024 at 18:12, Dale Scott ***@***.***> wrote:

> How about this as a solution - keep the master branch both MySQL *and*
> MariaDb compatible and create a new branch for optimising either MySQL or
> MariaDb depending which has the greatest value.
>
> Would it be possible to code a simple abstraction that would support
> either? It would make it easier to support the other dbms later with
> equivalent optimisations, or could possibly even be abandoned.
>
> Although my vote is still for MariaDb (and I think of many who get the
> choice), webERP is really for users and their hosting arrangements, and
> probably everyone else knows more about a real webERP user than me.
> However, imho it seems unlikely companies for which webERP is a good fit
> will eventually migrate to commercial Oracle. It seems more likely they
> will either be acquired, in which case they will be merged into an existing
> more advanced system, or their needs will continue to be met by webERP -
> the lowest cost best solution for their needs all things considered. Not
> considering the Oracle advantage, it seems "open source" and possibly
> performance are on the side of MariaDb, and commercial hosting on the side
> of MySQL.
>
> Is there a sense for how webERP is hosted?
>
> I checked Digital Ocean, who I have used in the past but am not currently.
> In addition to VPS they provide managed databases - but MySQL not MariaDb
> (I just checked). I'm not surprised as I cynically suspect Oracle may be
> compensating Digital Ocean to only offer MySQL (which would really be no
> different than many other cross-licensing agreements, although I have no
> reason to believe it).
>
>
> https://www.digitalocean.com/products/managed-databases?utm_campaign=search_us_en_brand&utm_adgroup=databases&utm_keyword=digitalocean%20dbaas&utm_matchtype=e&utm_adposition=&utm_creative=713874993326&utm_placement=&utm_device=c&utm_location=&utm_location=9001329&utm_term=digitalocean%20dbaas&utm_source=google&utm_medium=cpc&gad_source=1&gclid=Cj0KCQiAgJa6BhCOARIsAMiL7V93O55ficwotLkFgicycQuvEEu11kLesT3ChlZezhByzCHiqnL6ErcaAhZEEALw_wcB
>
> I assume Digital Ocean is not the only managed database provider. Is
> there  any reason why one would use a managed database with webERP? I guess
> if you had a Softaculus one-click install of webERP on a WHC VPS, and a
> one-click creation of a managed database on Digital Ocean, and you could
> configure webERP to use the managed database, you wouldn't need a systems
> administrator....
>
> I also checked WHC, one of my IP registrars and DNS hosting providers
> (ther other is No-IP). They also provide VPS (
> https://whc.ca/self-managed-cloud-servers) and support Softaculus for
> one-click installs (iiuc, I've never used nor understand Softaculus). The
> Softaculus library includes webERP 4.15.2 but neither MySQL or MariaDB, so
> I would have to ssh login to the VPS and install everything myself, as I
> suspect many webERP users would.
>
> @pakricard <https://github.com/pakricard> @efeoli
> <https://github.com/efeoli> who are your hosting service providers?
> @timschofield <https://github.com/timschofield> how are your clients
> hosting webERP?
>
> Thoughts?
>
> Fwiw, the Softaculus page for webERP 4.15.2 (
> https://softaculous.com/apps/erp) links to
> http://www.weberp.org/Support.html which doesn't exist.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/timschofield/webERP/discussions/238#discussioncomment-11386768>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAL6LC6EAWGE63QPWRSNAL2CS2XDAVCNFSM6AAAAABR2F5IMSVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMZYGY3TMOA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
www.weberp.org <http://www.weberpafrica.com>
@TimSchofield2
Blog: https://kwamoja.home.blog/

**Comment by pakricard** [comment]
Hi:

My current hosting company is Exabytes https://www.exabytes.com/, quite established in SE Asia. They offer both MariaDB (by default) and MySQL (on demand and with some cost involved, if I recall properly).
Plesk works OK for us, but we are not Unix savvy, so we enjoy a simple UI to manage most of the server issues, and ask the VPS supplier for “deep maintenance”.

As @timschofield mentioned, I agree the difference between MySQL and MariaDB (in this PERSISTENT case) is only during table creation, and it can be handled by ConnectDB_XXX.inc.

As per my experience, the gains in query performance are worth the effort.

**Comment by timschofield** [comment]
The initial effort wouldn't be so much as for the moment the only
difference would be these persistent fields. The main effort would be
changing the installer to accommodate it. If we can agree that this is
the way to go I am happy to code it.


On Wed, 27 Nov 2024 at 01:04, pakricard ***@***.***> wrote:
>
> Hi:
>
> My current hosting company is Exabytes https://www.exabytes.com/, quite established in SE Asia. They offer both MariaDB (by default) and MySQL (on demand and with some cost involved, if I recall properly).
> Plesk works OK for us, but we are not Unix savvy, so we enjoy a simple UI to manage most of the server issues, and ask the VPS supplier for “deep maintenance”.
>
> As @timschofield mentioned, I agree the difference between MySQL and MariaDB (in this PERSISTENT case) is only during table creation, and it can be handled by ConnectDB_XXX.inc.
>
> As per my experience, the gains in query performance are worth the effort.
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.Message ID: ***@***.***>



-- 
www.weberp.org
@TimSchofield2
Blog: https://kwamoja.home.blog/

**Comment by dalers** [comment]
@timschofield would the work needed dovetail into your existing plans for the installer?

@pakricard:
> gains in query performance are worth the effort.

Given a database roughly equivalent to @pakricard's, could the API be used to "do something" (likely many things in a loop) and measure how long the response took? Could the API be used to first _create_ a database of equivalent complexity to @pakricard's?

**Comment by timschofield** [comment]
Actually most of the work is done. I did this on KwaMoja. If everybody is
happy I can just do it.

Thanks
Tim

On Wed, 27 Nov 2024 at 18:05, Dale Scott ***@***.***> wrote:

> @timschofield <https://github.com/timschofield> would the work needed
> dovetail into your existing plans for the installer?
>
> @pakricard <https://github.com/pakricard>:
>
> gains in query performance are worth the effort.
>
> Given a database roughly equivalent to @pakricard
> <https://github.com/pakricard>'s, could the API be used to "do something"
> (likely many things in a loop) and measure how long the response took?
> Could the API be used to create a database of equivalent complexity used
> for testing?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/timschofield/webERP/discussions/238#discussioncomment-11397671>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAL6LDPEEUVCBISEFSWBYD2CYCVLAVCNFSM6AAAAABR2F5IMSVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMZZG43DOMI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
www.weberp.org <http://www.weberpafrica.com>
@TimSchofield2
Blog: https://kwamoja.home.blog/

**Reply by pakricard** [reply]
>  If everybody is happy I can just do it.

@timschofield  I am 100% OK with this approach, and thank you for the effort.

> Given a database roughly equivalent to @pakricard's, could the API be used to "do something" (likely many things in a loop) and measure how long the response took? Could the API be used to first create a database of equivalent complexity to @pakricard's?

@dalers : My DB has mainly the same complexity as standard webERP (I added some tables, fields here and there, but 95% of it is standard DB).  I mainly added warnings and complex business checks, that need heavy calculations queries.

It is probably a large DB (8,1 Gb as per Plesk calculations) with some large tables. 

In the past, I faced many timeout errors from the server due to heavy queries and these PERSISTENT fileds were a huge help, besides modifying or adding some specific indexes.
Probably some other webERP have larger DB's, not sure at all.

![image](https://github.com/user-attachments/assets/f8214e5d-99dc-4339-9fa9-a8a1b46aae7f)

I don't use the API at all.

BTW, I plan to donate all the "shareable code" once I'm up to date. Some of the improvements are a big "competitive advantage" in the retail business (daily optimization of stocks in every location, purchase planning, etc) and can't be shared as at least one competitor in Indonesia is using webERP, as they discovered we use it. Probably they will read this post ;-)

**Reply by dalers** [reply]
> I am 100% OK with this approach, and thank you for the effort.

You two know better than I. 

> My DB has mainly the same complexity as standard webERP

My bad wording. By complexity I meant the total number of items, customers, suppliers and transactions (purchases, sales, productions and adjustments). 

I've read through the API reference thinking perhaps I could write a script that loops through buying, production and selling on a weekly basic, and perhaps with some seasonal variations so the graphics have some nice cycles.

> as at least one competitor in Indonesia is using webERP

Imho, collaborative development (i.e. open source) works best when collaborators do not compete with each other directly. Competitors _could_ contribute equally so no one is at a disadvantage, but unlikely to happen in practise.

**Comment by timschofield** [comment]
OK guys, I have written these changes and pushed it to GitHub. The
installer will now give the option for mysqli or mariadb. If you are
currently using mariadb you need to change the line
$DBType = 'mysqli';
in config.php to
$DBType = 'mariadb';
I've tested the mysql version, but I don't have a mariadb installation to
hand so can someone test it for me?

Thanks
Tim

On Thu, 28 Nov 2024 at 18:01, Tim Schofield ***@***.***>
wrote:

> OK guys, I have written these changes and pushed it to GitHub. The
> installer will now give the option for mysqli or mariadb. If you are
> currently using mariadb you need to change the line
>
>
> On Thu, 28 Nov 2024 at 03:08, Dale Scott ***@***.***> wrote:
>
>> I am 100% OK with this approach, and thank you for the effort.
>>
>> You two know better than I.
>>
>> My DB has mainly the same complexity as standard webERP
>>
>> My bad wording. By complexity I meant the total number of items,
>> customers, suppliers and transactions (purchases, sales, productions and
>> adjustments).
>>
>> I've read through the API reference thinking perhaps I could write a
>> script that loops through buying, production and selling on a weekly basic,
>> and perhaps with some seasonal variations so the graphics have some nice
>> cycles.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/timschofield/webERP/discussions/238#discussioncomment-11401591>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAAL6LBRCNJZU5VIPEXFJXD2C2CLNAVCNFSM6AAAAABR2F5IMSVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCNBQGE2TSMI>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>
>
> --
> www.weberp.org <http://www.weberpafrica.com>
> @TimSchofield2
> Blog: https://kwamoja.home.blog/
>


-- 
www.weberp.org <http://www.weberpafrica.com>
@TimSchofield2
Blog: https://kwamoja.home.blog/

**Reply by efeoli** [reply]
Hi Tim,
I have just tested the updated installer, since I use maria.db. It did work very well for commit 8461. Then I imported the database from my current 4.15.2 setup and the update worked also quite well.

The only difference is that the uilities tab for Make a new Company, which is useful to me, now doesn't work, not even using a pre-installed database.

Enrique

**Comment by timschofield** [comment]
Hi Enrique, that script is next on my list to look at. Can you confirm the
update created the persistent fields?

Thanks
Tim

On Thu, 28 Nov 2024 at 21:17, efeoli ***@***.***> wrote:

> Hi Tim,
> I have just tested the updated installer, since I use maria.db. It did
> work very well for commit 8461. Then I imported the database from my
> current 4.15.2 setup and the update worked also quite well.
>
> The only difference is that the uilities tab for Make a new Company, which
> is useful to me, now doesn't work, not even using a pre-installed database.
>
> Enrique
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/timschofield/webERP/discussions/238#discussioncomment-11410269>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAL6LHXMJKG2AV32OXLV4L2C6B47AVCNFSM6AAAAABR2F5IMSVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCNBRGAZDMOI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
www.weberp.org <http://www.weberpafrica.com>
@TimSchofield2
Blog: https://kwamoja.home.blog/

**Reply by efeoli** [reply]
Since I am only a knowledgeable user, please tell me where to look for the persitent fields and inform thereafter.

**Comment by timschofield** [comment]
Hi Enrique, can you look for the table debtortrans for a field called
'balance', and let us know the definition for it?

Thanks,
Tim

On Thu, 28 Nov 2024 at 21:36, efeoli ***@***.***> wrote:
>
> Since I am only a knowledgeable user, please tell me where to look for the persitent fields and inform thereafter.
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.Message ID: ***@***.***>



-- 
www.weberp.org
@TimSchofield2
Blog: https://kwamoja.home.blog/

**Reply by efeoli** [reply]
Tim,
these are the fields of the updated 4.15.2 database. The field seems not present.

Should I search on a fresh install?
Enrique 
![Captura de pantalla_28-11-2024_162523_da4613 squidix net](https://github.com/user-attachments/assets/7d3693fc-d8b8-4a5f-a638-b36b681d08ac)

**Comment by timschofield** [comment]
Did the latest update that I committed today work?

Thanks
Tim

On Thu, 28 Nov 2024 at 22:36, efeoli ***@***.***> wrote:

> Tim,
> these are the fields of the updated 4.15.2 database. The field seems not
> present.
>
> Should I search on a fresh install?
> Enrique
> Captura.de.pantalla_28-11-2024_162523_da4613.squidix.net.jpeg (view on web)
> <https://github.com/user-attachments/assets/7d3693fc-d8b8-4a5f-a638-b36b681d08ac>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/timschofield/webERP/discussions/238#discussioncomment-11410590>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAL6LCQ4CC2XSYEMEVPKE32C6LGTAVCNFSM6AAAAABR2F5IMSVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCNBRGA2TSMA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
www.weberp.org <http://www.weberpafrica.com>
@TimSchofield2
Blog: https://kwamoja.home.blog/

**Reply by efeoli** [reply]
Yes, I used commit 8461, the last one committed and both the fresh install (choosing Maria.db) and subsequent update of the 4.15.2 database, went on very smoothly.
Enrique

**Reply by efeoli** [reply]
Oh, by the way, it wasn’t necessary to change db preference in the config.php file. It was already chosen by the installer.

**Comment by timschofield** [comment]
Yes, with a fresh install it doesn't need it, just an upgrade.

On Thu, 28 Nov 2024 at 22:54, efeoli ***@***.***> wrote:
>
> Oh, by the way, it wasn’t necessary to change db preference in the config.php file. It was already chosen by the installer.
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.Message ID: ***@***.***>



-- 
www.weberp.org
@TimSchofield2
Blog: https://kwamoja.home.blog/

**Reply by efeoli** [reply]
Hi All,
I have experienced a boost in browsing speed with specific MariaDB connection.

Effort was quite worth it and proponents of the enhancement were right. Thanks.
Enrique

**Reply by pakricard** [reply]
Hi @efeoli:
The real improvement will come once we change all instances of the AS expressions for the calculated fields, as then queries will not be that calculation-heavy, as all persistent fields are "pre-calculated". In addition, we can set indexes with these fields.

**Reply by efeoli** [reply]
Hi @pakricard,
thank you for explaining, improvement which sounds exciting. I will test again when completed.
Enrique

**Comment by pakricard** [comment]
Tim:
I tested the ConnectDB_mariadb.inc in my installation and worked OK.
Just found that config.distrib.php was using $host instead of $Host. I commited the fix. 
Thanks!

**Comment by dalers** [comment]
@timschofield upgrade of existing db to mariadb was a non-event. Thanks!

@pakricard thanks for fixing this, I noticed the difference but didn't know if it was an error and which one was correct.

**Comment by timschofield** [comment]
Hi Ricard/Enrique

Anyone fancy the job of tracking down these instances? :)

Tim



On Fri, 29 Nov 2024 at 23:23, efeoli ***@***.***> wrote:

> Hi @pakricard <https://github.com/pakricard>,
> thank you for explaining, improvement which sounds exciting. I will test
> again when completed.
> Enrique
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/timschofield/webERP/discussions/238#discussioncomment-11419473>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAL6LEY4JPQGCHQMFEFX432DDZPFAVCNFSM6AAAAABR2F5IMSVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCNBRHE2DOMY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
www.weberp.org <http://www.weberpafrica.com>
@TimSchofield2
Blog: https://kwamoja.home.blog/

**Reply by efeoli** [reply]
Tim,
of matters affecting common users, still stock images appear thumbnail size and link to wiki has warnings not present previously in my setting.
Enrique 
![Captura de pantalla_2-12-2024_94850_kwa bioacyl com](https://github.com/user-attachments/assets/d947d0db-f1aa-4e51-b063-26d88f8291af)

**Reply by timschofield** [reply]
Hi Enrique,

If you go to the demo and SelectProduct.php, and then choose the part *3_3 you should see a thumbnail image moving the mouse over the image it will expand.

![item_image1](https://github.com/user-attachments/assets/d61308a1-b5b4-4503-be0b-4fb51fafac36)
![item_image2](https://github.com/user-attachments/assets/cba49cde-7b70-49f5-9e16-efbe825e1b29)

Is this not what happens on your system?

Thanks
Tim

**Reply by efeoli** [reply]
Hi Tim,
sorry I replied for issue 250 in this discussion.

My testing was done on the updated version with prior production data. With latest changes I cannot compare with the demo as I used to do earlier. Both versions mounted with Make New Company utility.

Once you load data to simulate a production environment, it seems not to work as you show.
Enrique

**Comment by dalers** [comment]
@timschofield after selecting an item in SelectProduct.php, the thumbnail does not expand when I hover over it.

Is there a minimum size in pixels for the original image for the "expand" to work? Does it matter if the original image is JPG or PNG?

![image](https://github.com/user-attachments/assets/189769bb-eab7-4a9c-b834-1048219f43ac)

Fwiw, latest code, PHP 8.1 on server, error_reporting(-1) and $Debug = 0 in config.php, no errors in Apache error log.

**Comment by timschofield** [comment]
What is view part number *3-3 in the demo. Do you see the same as was in
those screenshots I published earlier.

Thanks
Tim

On Mon, 2 Dec 2024 at 19:57, Dale Scott ***@***.***> wrote:

> @timschofield <https://github.com/timschofield> after selecting an item
> in SelectProduct.php, the thumbnail does not expand when I hover over it.
>
> Is there a minimum size in pixels for the original image for the "expand"
> to work? Does it matter if the original image is JPG or PNG?
>
> image.png (view on web)
> <https://github.com/user-attachments/assets/189769bb-eab7-4a9c-b834-1048219f43ac>
>
> Fwiw, latest code, PHP 8.1 on server, error_reporting(-1) and $Debug = 0
> in config.php, no errors in Apache error log.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/timschofield/webERP/discussions/238#discussioncomment-11440664>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAL6LF2ZN2T6HXUD7FG2XD2DS3RNAVCNFSM6AAAAABR2F5IMSVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCNBUGA3DMNA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
www.weberp.org <http://www.weberpafrica.com>
@TimSchofield2
Blog: https://kwamoja.home.blog/

**Comment by dalers** [comment]
DIscussion of image visibility, image "expand" and image size continued in Issue #250. Please do not add additional comments here.

**Comment by pakricard** [comment]
Hi Tim:

Yes, I can do the search and replace of the SQL queries using persistent fields. Done it already in my install 2 years ago. I'll add it to my to-do list.

**Comment by pakricard** [comment]
Done and dusted!
